In [1]:
from __future__ import print_function
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms

In [11]:
!git clone https://github.com/ankited3/S6.git

Cloning into 'S6'...
remote: Enumerating objects: 19, done.
remote: Counting objects: 100% (19/19), done.
remote: Compressing objects: 100% (10/10), done.
remote: Total 19 (delta 9), reused 14 (delta 7), pack-reused 0 (from 0)
Receiving objects: 100% (19/19), 8.86 KiB | 168.00 KiB/s, done.
Resolving deltas: 100% (9/9), done.


In [8]:
# CUDA?
cuda = torch.cuda.is_available()
print("CUDA Available?", cuda)

CUDA Available? True


In [12]:
%cd S6

/content/S6/S6/S6


In [4]:
# Assuming your master file is 'master_script.py' inside the cloned folder
%run model_1.py

/content/S6


100%|██████████| 9.91M/9.91M [00:00<00:00, 18.6MB/s]
100%|██████████| 28.9k/28.9k [00:00<00:00, 495kB/s]
100%|██████████| 1.65M/1.65M [00:00<00:00, 4.58MB/s]
100%|██████████| 4.54k/4.54k [00:00<00:00, 14.2MB/s]
/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py:627: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  warnings.warn(


CUDA Available? True
cuda
----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
            Conv2d-1           [-1, 10, 26, 26]              90
              ReLU-2           [-1, 10, 26, 26]               0
            Conv2d-3           [-1, 16, 24, 24]           1,440
              ReLU-4           [-1, 16, 24, 24]               0
            Conv2d-5           [-1, 20, 22, 22]           2,880
              ReLU-6           [-1, 20, 22, 22]               0
         MaxPool2d-7           [-1, 20, 11, 11]               0
            Conv2d-8           [-1, 10, 11, 11]             200
            Conv2d-9             [-1, 16, 9, 9]           1,440
             ReLU-10             [-1, 16, 9, 9]               0
           Conv2d-11             [-1, 20, 7, 7]           2,880
             ReLU-12             [-1, 20, 7, 7]               0
           Conv2d-13             [-1, 10, 7, 7]             200
             

Loss=0.37065640091896057 Batch_id=468 Accuracy=38.66: 100%|██████████| 469/469 [00:13<00:00, 33.86it/s]



Test set: Average loss: 0.3754, Accuracy: 8871/10000 (88.71%)

Model 1 EPOCH: 1


Loss=0.14253751933574677 Batch_id=468 Accuracy=92.66: 100%|██████████| 469/469 [00:13<00:00, 35.34it/s]



Test set: Average loss: 0.1245, Accuracy: 9599/10000 (95.99%)

Model 1 EPOCH: 2


Loss=0.1260908842086792 Batch_id=468 Accuracy=96.71: 100%|██████████| 469/469 [00:13<00:00, 34.40it/s]



Test set: Average loss: 0.0788, Accuracy: 9756/10000 (97.56%)

Model 1 EPOCH: 3


Loss=0.1512085348367691 Batch_id=468 Accuracy=97.61: 100%|██████████| 469/469 [00:13<00:00, 35.22it/s]



Test set: Average loss: 0.0747, Accuracy: 9764/10000 (97.64%)

Model 1 EPOCH: 4


Loss=0.09376107901334763 Batch_id=468 Accuracy=98.01: 100%|██████████| 469/469 [00:13<00:00, 35.36it/s]



Test set: Average loss: 0.0581, Accuracy: 9810/10000 (98.10%)

Model 1 EPOCH: 5


Loss=0.0836348906159401 Batch_id=468 Accuracy=98.34: 100%|██████████| 469/469 [00:13<00:00, 35.62it/s]



Test set: Average loss: 0.0507, Accuracy: 9826/10000 (98.26%)

Model 1 EPOCH: 6


Loss=0.051443781703710556 Batch_id=468 Accuracy=98.48: 100%|██████████| 469/469 [00:13<00:00, 35.15it/s]



Test set: Average loss: 0.0416, Accuracy: 9863/10000 (98.63%)

Model 1 EPOCH: 7


Loss=0.03270573168992996 Batch_id=468 Accuracy=98.59: 100%|██████████| 469/469 [00:13<00:00, 35.82it/s]



Test set: Average loss: 0.0413, Accuracy: 9865/10000 (98.65%)

Model 1 EPOCH: 8


Loss=0.06622251868247986 Batch_id=468 Accuracy=98.72: 100%|██████████| 469/469 [00:13<00:00, 35.12it/s]



Test set: Average loss: 0.0427, Accuracy: 9860/10000 (98.60%)

Model 1 EPOCH: 9


Loss=0.008682029321789742 Batch_id=468 Accuracy=98.85: 100%|██████████| 469/469 [00:13<00:00, 35.82it/s]



Test set: Average loss: 0.0403, Accuracy: 9864/10000 (98.64%)

Model 1 EPOCH: 10


Loss=0.04237562417984009 Batch_id=468 Accuracy=98.89: 100%|██████████| 469/469 [00:13<00:00, 35.68it/s]



Test set: Average loss: 0.0426, Accuracy: 9866/10000 (98.66%)

Model 1 EPOCH: 11


Loss=0.023287305608391762 Batch_id=468 Accuracy=98.93: 100%|██████████| 469/469 [00:12<00:00, 36.33it/s]



Test set: Average loss: 0.0438, Accuracy: 9857/10000 (98.57%)

Model 1 EPOCH: 12


Loss=0.08114940673112869 Batch_id=468 Accuracy=99.06: 100%|██████████| 469/469 [00:13<00:00, 34.46it/s]



Test set: Average loss: 0.0347, Accuracy: 9882/10000 (98.82%)

Model 1 EPOCH: 13


Loss=0.02922806888818741 Batch_id=468 Accuracy=99.08: 100%|██████████| 469/469 [00:13<00:00, 35.58it/s]



Test set: Average loss: 0.0410, Accuracy: 9860/10000 (98.60%)

Model 1 EPOCH: 14


Loss=0.008141984231770039 Batch_id=468 Accuracy=99.13: 100%|██████████| 469/469 [00:13<00:00, 34.92it/s]



Test set: Average loss: 0.0383, Accuracy: 9883/10000 (98.83%)



#Model 1
##Target
To have a model with basic code setup, transform, data loader, training and test loop
##Result
1. Parameters:14030
2. Best Training Accuracy:99.13
3. Best Test Accuracy:98.83
##Analysis
Model is good but number of parameter is more than 8k need to reduce parameters
Model is slight overfitting but can be pushed further


In [5]:
%run model_2.py

CUDA Available? True


/usr/local/lib/python3.12/dist-packages/torchvision/datasets/mnist.py:76: UserWarning: train_data has been renamed data
  warnings.warn("train_data has been renamed data")


cuda
----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
            Conv2d-1            [-1, 8, 26, 26]              72
              ReLU-2            [-1, 8, 26, 26]               0
       BatchNorm2d-3            [-1, 8, 26, 26]              16
           Dropout-4            [-1, 8, 26, 26]               0
            Conv2d-5           [-1, 12, 24, 24]             864
              ReLU-6           [-1, 12, 24, 24]               0
       BatchNorm2d-7           [-1, 12, 24, 24]              24
           Dropout-8           [-1, 12, 24, 24]               0
            Conv2d-9           [-1, 16, 24, 24]             192
        MaxPool2d-10           [-1, 16, 12, 12]               0
           Conv2d-11           [-1, 12, 10, 10]           1,728
             ReLU-12           [-1, 12, 10, 10]               0
      BatchNorm2d-13           [-1, 12, 10, 10]              24
          Dropout-14           [-1

Loss=0.06956028938293457 Batch_id=468 Accuracy=89.40: 100%|██████████| 469/469 [00:13<00:00, 35.08it/s]



Test set: Average loss: 0.0701, Accuracy: 9805/10000 (98.05%)

Model 2 EPOCH: 1


Loss=0.05645255371928215 Batch_id=468 Accuracy=98.01: 100%|██████████| 469/469 [00:13<00:00, 35.54it/s]



Test set: Average loss: 0.0529, Accuracy: 9839/10000 (98.39%)

Model 2 EPOCH: 2


Loss=0.025395462289452553 Batch_id=468 Accuracy=98.52: 100%|██████████| 469/469 [00:13<00:00, 35.23it/s]



Test set: Average loss: 0.0377, Accuracy: 9880/10000 (98.80%)

Model 2 EPOCH: 3


Loss=0.09736192971467972 Batch_id=468 Accuracy=98.71: 100%|██████████| 469/469 [00:13<00:00, 35.78it/s]



Test set: Average loss: 0.0366, Accuracy: 9876/10000 (98.76%)

Model 2 EPOCH: 4


Loss=0.010401505045592785 Batch_id=468 Accuracy=98.83: 100%|██████████| 469/469 [00:13<00:00, 35.26it/s]



Test set: Average loss: 0.0376, Accuracy: 9878/10000 (98.78%)

Model 2 EPOCH: 5


Loss=0.026180721819400787 Batch_id=468 Accuracy=98.92: 100%|██████████| 469/469 [00:12<00:00, 36.12it/s]



Test set: Average loss: 0.0313, Accuracy: 9901/10000 (99.01%)

Model 2 EPOCH: 6


Loss=0.052813220769166946 Batch_id=468 Accuracy=99.01: 100%|██████████| 469/469 [00:13<00:00, 35.29it/s]



Test set: Average loss: 0.0274, Accuracy: 9903/10000 (99.03%)

Model 2 EPOCH: 7


Loss=0.030792340636253357 Batch_id=468 Accuracy=99.09: 100%|██████████| 469/469 [00:13<00:00, 35.64it/s]



Test set: Average loss: 0.0325, Accuracy: 9896/10000 (98.96%)

Model 2 EPOCH: 8


Loss=0.014338679611682892 Batch_id=468 Accuracy=99.14: 100%|██████████| 469/469 [00:13<00:00, 35.44it/s]



Test set: Average loss: 0.0245, Accuracy: 9919/10000 (99.19%)

Model 2 EPOCH: 9


Loss=0.008646237663924694 Batch_id=468 Accuracy=99.23: 100%|██████████| 469/469 [00:13<00:00, 36.04it/s]



Test set: Average loss: 0.0227, Accuracy: 9926/10000 (99.26%)

Model 2 EPOCH: 10


Loss=0.04673122987151146 Batch_id=468 Accuracy=99.23: 100%|██████████| 469/469 [00:13<00:00, 35.09it/s]



Test set: Average loss: 0.0237, Accuracy: 9913/10000 (99.13%)

Model 2 EPOCH: 11


Loss=0.019146980717778206 Batch_id=468 Accuracy=99.30: 100%|██████████| 469/469 [00:13<00:00, 35.67it/s]



Test set: Average loss: 0.0244, Accuracy: 9923/10000 (99.23%)

Model 2 EPOCH: 12


Loss=0.05278274044394493 Batch_id=468 Accuracy=99.29: 100%|██████████| 469/469 [00:13<00:00, 35.55it/s]



Test set: Average loss: 0.0246, Accuracy: 9917/10000 (99.17%)

Model 2 EPOCH: 13


Loss=0.08865227550268173 Batch_id=468 Accuracy=99.34: 100%|██████████| 469/469 [00:13<00:00, 35.30it/s]



Test set: Average loss: 0.0264, Accuracy: 9911/10000 (99.11%)

Model 2 EPOCH: 14


Loss=0.023110099136829376 Batch_id=468 Accuracy=99.35: 100%|██████████| 469/469 [00:13<00:00, 35.37it/s]



Test set: Average loss: 0.0239, Accuracy: 9922/10000 (99.22%)



#Model 2
##Target
To have a model similar to model 1 but with less than 8k parameters by using GAP layer, batch normalization so that model parameter can be reduced, need to reduce over fitting by adding drop out
##Result
1. Parameters:7632
2. Best Training Accuracy:99.35
3. Best Test Accuracy:99.23
##Analysis
1. Model is better and number of parameter are under 8k
2. Model is overfitting but overfitting is lesser compare to model 1
3. Model is still not touching 99.4 test accuracy



In [13]:
%run model_3.py

100%|██████████| 9.91M/9.91M [00:00<00:00, 18.6MB/s]
100%|██████████| 28.9k/28.9k [00:00<00:00, 497kB/s]
100%|██████████| 1.65M/1.65M [00:00<00:00, 4.60MB/s]
100%|██████████| 4.54k/4.54k [00:00<00:00, 12.1MB/s]


CUDA Available? True
----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
            Conv2d-1            [-1, 8, 26, 26]              72
              ReLU-2            [-1, 8, 26, 26]               0
       BatchNorm2d-3            [-1, 8, 26, 26]              16
           Dropout-4            [-1, 8, 26, 26]               0
            Conv2d-5           [-1, 12, 24, 24]             864
              ReLU-6           [-1, 12, 24, 24]               0
       BatchNorm2d-7           [-1, 12, 24, 24]              24
           Dropout-8           [-1, 12, 24, 24]               0
            Conv2d-9           [-1, 16, 24, 24]             192
        MaxPool2d-10           [-1, 16, 12, 12]               0
           Conv2d-11           [-1, 12, 10, 10]           1,728
             ReLU-12           [-1, 12, 10, 10]               0
      BatchNorm2d-13           [-1, 12, 10, 10]              24
          Dropout-

Train: Loss=0.0761 Batch_id=937 Accuracy=93.76: 100%|██████████| 938/938 [00:23<00:00, 40.45it/s]



Test set: Average loss: 0.0009, Accuracy: 9801/10000 (98.01%)

Model 3 EPOCH: 1


Train: Loss=0.0386 Batch_id=937 Accuracy=97.53: 100%|██████████| 938/938 [00:22<00:00, 41.24it/s]



Test set: Average loss: 0.0006, Accuracy: 9878/10000 (98.78%)

Model 3 EPOCH: 2


Train: Loss=0.0064 Batch_id=937 Accuracy=98.00: 100%|██████████| 938/938 [00:22<00:00, 41.25it/s]



Test set: Average loss: 0.0004, Accuracy: 9908/10000 (99.08%)

Model 3 EPOCH: 3


Train: Loss=0.0899 Batch_id=937 Accuracy=98.14: 100%|██████████| 938/938 [00:22<00:00, 41.10it/s]



Test set: Average loss: 0.0007, Accuracy: 9854/10000 (98.54%)

Model 3 EPOCH: 4


Train: Loss=0.0023 Batch_id=937 Accuracy=98.90: 100%|██████████| 938/938 [00:22<00:00, 40.89it/s]



Test set: Average loss: 0.0003, Accuracy: 9932/10000 (99.32%)

Model 3 EPOCH: 5


Train: Loss=0.0628 Batch_id=937 Accuracy=98.98: 100%|██████████| 938/938 [00:22<00:00, 41.37it/s]



Test set: Average loss: 0.0003, Accuracy: 9937/10000 (99.37%)

Model 3 EPOCH: 6


Train: Loss=0.0021 Batch_id=937 Accuracy=99.03: 100%|██████████| 938/938 [00:24<00:00, 38.38it/s]



Test set: Average loss: 0.0003, Accuracy: 9935/10000 (99.35%)

Model 3 EPOCH: 7


Train: Loss=0.0300 Batch_id=937 Accuracy=99.10: 100%|██████████| 938/938 [00:23<00:00, 40.16it/s]



Test set: Average loss: 0.0003, Accuracy: 9944/10000 (99.44%)

Model 3 EPOCH: 8


Train: Loss=0.0224 Batch_id=937 Accuracy=99.12: 100%|██████████| 938/938 [00:23<00:00, 39.85it/s]



Test set: Average loss: 0.0003, Accuracy: 9943/10000 (99.43%)

Model 3 EPOCH: 9


Train: Loss=0.0026 Batch_id=937 Accuracy=99.21: 100%|██████████| 938/938 [00:23<00:00, 40.63it/s]



Test set: Average loss: 0.0003, Accuracy: 9946/10000 (99.46%)

Model 3 EPOCH: 10


Train: Loss=0.0617 Batch_id=937 Accuracy=99.22: 100%|██████████| 938/938 [00:23<00:00, 40.51it/s]



Test set: Average loss: 0.0003, Accuracy: 9945/10000 (99.45%)

Model 3 EPOCH: 11


Train: Loss=0.0122 Batch_id=937 Accuracy=99.21: 100%|██████████| 938/938 [00:23<00:00, 40.60it/s]



Test set: Average loss: 0.0003, Accuracy: 9939/10000 (99.39%)

Model 3 EPOCH: 12


Train: Loss=0.0360 Batch_id=937 Accuracy=99.25: 100%|██████████| 938/938 [00:22<00:00, 41.70it/s]



Test set: Average loss: 0.0003, Accuracy: 9940/10000 (99.40%)

Model 3 EPOCH: 13


Train: Loss=0.0376 Batch_id=937 Accuracy=99.21: 100%|██████████| 938/938 [00:22<00:00, 41.37it/s]



Test set: Average loss: 0.0003, Accuracy: 9944/10000 (99.44%)

Model 3 EPOCH: 14


Train: Loss=0.0030 Batch_id=937 Accuracy=99.24: 100%|██████████| 938/938 [00:22<00:00, 41.60it/s]



Test set: Average loss: 0.0003, Accuracy: 9941/10000 (99.41%)



#Model 3
##Target
1. To have a model similar to model 2 but with higher test accuracy consistenly above 99.4%,
2. Need to train better as last model is still overfitting with the help of image augmentation and will be playing naively with Adam optimizer and Scheduler to improve the accuracy
##Result
1. Parameters:7632
2. Best Training Accuracy:99.46
3. Best Test Accuracy:99.24
##Analysis
1. Model is quite good and number of parameter are under 8k
2. Model is underfitting as training accuracy is less than test accuracy
3. Model is consistently above 99.4% after epoch 9

